### PR TITLE
Handle comments in match better (fixes #344)

### LINKED ISSUE #344: Comments on last match arm ignored
```
fn main() {
    match foo {
        Bar => Baz,
        Qux => Quux,
        _ => foo, // foobar
    }
}
```

The last comment gets removed.

### PR DESCRIPTION
**Pull request by Manishearth**
This unfortunately replaces block comments with a block of line comments,
but I don't know a better way of doing this without rewriting `rewrite_comment` to have
a mode for preserving comment styles

r? @nrc

### CODE REVIEW

**Comment by Manishearth** [comment]
Actually, we could be smarter here. There are three situations which we club together:

```rust
match _ {
  a => {}  // comment pertaining to arm a
  b => {}
}

match _ {
  a => {}
   // comment pertaining to arm b (might be multiline)
  b => {}
}

match _ {
  a => {} // comment pertaining to line a
   // comment pertaining to arm b (might be multiline)
  b => {}
}
```

And, of course, block comments (which we should just feed to `rewrite_comment`, I guess). Line comments at the end of an arm should be moved above the arm, I guess.

**Comment by nrc** [comment]
Could you add some tests please.

File: src/expr.rs [hunk 1]
```diff
@@ -666,7 +667,8 @@ fn rewrite_match(context: &RewriteContext,
         if !missed_str.is_empty() {
             result.push('\n');
             result.push_str(&arm_indent_str);
-            result.push_str(missed_str);
+            result.push_str(&rewrite_comment(&missed_str, false,
+                                             width, arm_indent, context.config));
```
Comment (nrc) [inline src/expr.rs]: Nit: one arg per line

Slightly worried that we don't reformat this in cargo test.

File: src/expr.rs [hunk 2]
```diff
@@ -682,11 +684,23 @@ fn rewrite_match(context: &RewriteContext,
...
+    let last_comment = context.snippet(mk_sp(arm_end_pos(&arms[arms.len() - 1]), span.hi));
+    let last_comment = match last_comment.find_uncommented(",") {
+        Some(n) => &last_comment[n+1..],
+        None => &last_comment[..],
+    };
+    let last_comment = match last_comment.find_uncommented("}") {
+        Some(n) => &last_comment[..n-1],
+        None => &last_comment[..],
+    };
+    let last_comment = last_comment.trim();
```
Comment (nrc) [inline src/expr.rs]: I feel like this chunk of code is repeated at least once elsewhere in the codebase and ought to be factored out into a function if possible.

**Comment by nrc** [comment]
With functions, we use some heuristics to try and distinguish between pre-comments and post-comments, although I think that depends on the commas. We could special case empty arms, because that seems like the only reasonable place to get a post-comment for match arms (well, also the comma case). I think we can do better with comments that way.

I'd be happy to land this patch as is (well, with the nits addressed) and do better comments later.

File: src/expr.rs (hunk 2 above)
Comment (Manishearth) [inline src/expr.rs]: Can't find anything :/

**Comment by Manishearth** [comment]
Updated.

**Comment by nrc** [comment]
I think you reformatted the wrong expr

**Comment by Manishearth** [comment]
done

**Comment by nrc** [comment]
Thanks!

**Pull request merged**